chain/chain/src/chain.rs

```diff
@@ -1390,19 +1391,36 @@ impl Chain {
         let shard_state_header = self.get_received_state_header(shard_id, sync_hash)?;
         let ShardStateSyncResponseHeader { chunk, .. } = shard_state_header;
         let state_root = chunk.header.inner.prev_state_root;
-        if !self.runtime_adapter.validate_state_part(&state_root, part_id, num_parts, data) {
-            byzantine_assert!(false);
-            return Err(ErrorKind::Other(
-                "set_state_part failed: validate_state_part failed".into(),
-            )
-            .into());
+        match lzma::decompress(data) {
+            Ok(decompressed_data) => {
+                if !self.runtime_adapter.validate_state_part(
+                    &state_root,
+                    part_id,
+                    num_parts,
+                    &decompressed_data,
+                ) {
+                    byzantine_assert!(false);
+                    return Err(ErrorKind::Other(
+                        "set_state_part failed: validate_state_part failed".into(),
+                    )
+                    .into());
+                }
+
+                // Saving the part data.
+                let mut store_update = self.store.owned_store().store_update();
+                let key = StatePartKey(sync_hash, shard_id, part_id).try_to_vec()?;
+                store_update.set_ser(COL_STATE_PARTS, &key, &decompressed_data)?;
+                store_update.commit()?;
+            }
+            _ => {
+                byzantine_assert!(false);
+                return Err(ErrorKind::Other(
+                    "set_state_part failed: lzma::decompress failed".into(),
```

Comment:
Why not show the decompression error here? Might help with debugging

Reply:
It should not fail at all.

Reply:
So every byte sequence can be decoded?

Reply:
Not failing means that failing caused by clearly malicious data.